### PR TITLE
Feature/json out

### PR DESCRIPTION
**Pull request by asabya**

### CODE REVIEW

File: lib/ss_light.go [hunk 1]
```diff
@@ -382,18 +386,18 @@ func (l *LightClient) Start(
 
 	_, err = io.Copy(dst, rsc)
 	if err != nil {
-		return "Failed writing to destination", err
+		return NewOut(500, "Failed writing to destination", err.Error(), nil)
 	}
-
+	downloadTime := startTime - time.Now().Unix()
```
Comment (aloknerurkar) [inline lib/ss_light.go]: This should be time.Now() - startTime right?

File: examples/litepeer/litepeer.go [hunk 2]
```diff
@@ -77,38 +78,36 @@ func main() {
...
-	if u.started {
-		CallClear()
-	} else {
-		u.started = true
-	}
-	fmt.Printf("Progress %d%%\n", p)
+func (u *updateProgress) UpdateProgress(p, downloadedSize, fullSize int) {
+	dMB := float32(downloadedSize) / (1024 * 1024)
+	fMB := float32(fullSize) / (1024 * 1024)
+	out := lib.NewOut(200, "Progress", "", fmt.Sprintf("%d%% (%.2fMB / %.2fMB)", p, dMB, fMB))
```
Comment (aloknerurkar) [inline examples/litepeer/litepeer.go:167]: Everything is INT here... why are we using 0.2f format specifier?

File: examples/litepeer/litepeer.go [hunk 3]
```diff
@@ -20,6 +20,7 @@ var (
 	stat        = flag.Bool("stat", false, "Get stat of the last fetch")
 	enableLog   = flag.Bool("logToStderr", false, "Enable app logs on stderr")
 	showProg    = flag.Bool("progress", false, "Enable progress on stdout")
+	jsonOut     = flag.Bool("json", false, "Display output in json format")
 )
 
 func returnError(err string, printUsage bool) {
```
Comment (aloknerurkar) [inline examples/litepeer/litepeer.go:27]: We need a man page type usage doc.... we can use this same output to display on the CLI docs web

File: examples/litepeer/litepeer.go (hunk 2 above)
Comment (asabya) [reply examples/litepeer/litepeer.go:167]:
```
dMB := float32(downloadedSize) / (1024 * 1024)
	fMB := float32(fullSize) / (1024 * 1024)
```

using `%.2f` after converting bytes into mb

File: lib/ss_light.go (hunk 1 above)
Comment (asabya) [reply lib/ss_light.go]: yes, you are right

File: examples/litepeer/litepeer.go (hunk 3 above)
Comment (asabya) [reply examples/litepeer/litepeer.go:27]: We can have a `-help` flag to `PrintDefaults` usage

File: examples/litepeer/litepeer.go [hunk 4]
```diff
@@ -13,27 +13,79 @@ import (
 
 // Command arguments
 var (
-	destination = flag.String("dst", ".", "Path to store downloaded file")
+	destination = flag.String("dst", ".", "Filename to store downloaded file")
```
Comment (aloknerurkar) [inline examples/litepeer/litepeer.go]: "Complete file path on disk to store downloaded file"

File: lib/ss_light.go [hunk 5]
```diff
@@ -262,24 +274,27 @@ func (l *LightClient) Start(
...
 		ipfslite.Libp2pOptionsExtra...,
 	)
 	if err != nil {
 		log.Errorf("Failed setting up libp2p node Err: %s", err.Error())
-		return "Failed setting up p2p peer", err
+		return NewOut(internalError, "Failed setting up p2p peer", err.Error(), nil)
 	}
 
+	// STEP : Host Created
+	showStep(success, "Host created", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: This is not required..... Customer may might not understand what Host means

File: lib/ss_light.go [hunk 6]
```diff
@@ -301,6 +323,9 @@ func (l *LightClient) Start(
 				case <-ctx.Done():
 					return
 				case <-time.After(time.Second * 30):
+					// STEP : Re-fetching metadata
+					showStep(success, "Re-fetching metadata", l.jsonOut)
+
 					// Allow 30 seconds to see if new leaders were added
 					newMtdt, err := getInfo(sharable, metadata.Cookie.Id, l.pubKey)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: Lets remove the refetching now. This was required due to the previous flow.... 

Now we should just retry connecting to the same list if count < len(metadata.Cookie.Leaders)

File: lib/ss_light.go [hunk 7]
```diff
@@ -383,18 +416,26 @@ func (l *LightClient) Start(
...
 	if err != nil {
-		return "Failed writing to destination", err
+		return NewOut(internalError, "Failed writing to destination", err.Error(), nil)
 	}
+	downloadTime := time.Now().Unix() - startTime
+	// STEP : Download Finished
+	showStep(success, "Finished download", l.jsonOut)
 
+	// STEP : Waiting for micropayments clean up
+	showStep(success, "Sending micropayments", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: Make it "Finishing download"
Instead of "Finished" + "Sending"

File: lib/ss_light.go [hunk 8]
```diff
@@ -289,10 +300,17 @@ func (l *LightClient) Start(
 	lite, err := ipfslite.New(ctx, l.ds, h, dht, cfg)
 	if err != nil {
 		log.Errorf("Failed setting up p2p xfer node Err: %s", err.Error())
-		return "Failed setting up light client", err
+		return NewOut(internalError, "Failed setting up light client", err.Error(), nil)
 	}
 
+	// STEP : Download agent created
+	showStep(success, "Download agent created", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: "initialized" instead of "created" makes more sense

File: lib/ss_light.go [hunk 9]
```diff
@@ -289,10 +300,17 @@ func (l *LightClient) Start(
...
+
 	count := lite.Bootstrap(metadata.Cookie.Leaders)
+
+	// STEP : Bootstrap done
+	showStep(success, "Bootstrapped", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: "Bootstrapped agent"

File: lib/ss_light.go [hunk 10]
```diff
@@ -301,29 +319,17 @@ func (l *LightClient) Start(
...
-						count += lite.Bootstrap(newLeadersToAdd)
-					}
 					if time.Since(start) > time.Minute*15 {
 						log.Warn("Tried getting more peers for 15mins")
+						showStep(timeoutError, "Download timed out", l.jsonOut)
 						return
 					}
+					// Try to re-bootstrap if client was unable to bootstrap previously
+					if count < len(metadata.Cookie.Leaders) {
+						count += lite.Bootstrap(metadata.Cookie.Leaders)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: This doesnt seem right. If you connect to 2/5 nodes. But then you try to connect to 5 again. Say you only connect to 2 again but its the same 2 nodes. In which case your count will not be correct.

This needs to be count = bootstrap()

File: lib/ss_light.go [hunk 11]
```diff
@@ -301,29 +319,17 @@ func (l *LightClient) Start(
...
+						// STEP : Re-Bootstrap done
+						showStep(success, "Re-Bootstrapped", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: Only if we are able to connect to more peers we should fire an update.

So
```
oldCount := count
count = lite.Bootstrap(metadata.Cookie.Leaders)
if count > oldCount {
    showStep(success, "Found more peers to connect", l.jsonOut)
}
```

File: lib/ss_light.go [hunk 12]
```diff
@@ -383,18 +393,24 @@ func (l *LightClient) Start(
...
 	// Wait 5 secs for SCP to send all MPs. This can be optimized
 	<-time.After(time.Second * 5)
 
-	err = updateInfo(metadata)
+	err = updateInfo(metadata, downloadTime)
 	if err != nil {
 		log.Warn("Failed updating metadata after download Err: %s", err.Error())
 	}
+	// STEP : Updated Cookie
+	showStep(success, "Updating cookie", l.jsonOut)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: Remove this. This is sent after Cookie is updated. Cookie doesn't mean anything to the user of this.

File: lib/ss_light.go [hunk 13]
```diff
@@ -229,31 +237,35 @@ func NewLightClient(
...
-	dst, err := os.Create(
-		combineArgs(fpSeparator, l.destination, metadata.Cookie.Filename))
+	if l.destination == "." {
+		l.destination = combineArgs(fpSeparator, l.destination, metadata.Cookie.Filename)
+	}
+	dst, err := os.Create(l.destination)
 	if err != nil {
 		log.Errorf("Failed creating dest file Err: %s", err.Error())
-		return "Failed creating destination file", err
+		return NewOut(internalError, "Failed creating destination file", err.Error(), nil)
```
Comment (aloknerurkar) [inline lib/ss_light.go]: This is a different type of error. This could only happen if user gave destination path incorrect

**Review by aloknerurkar — APPROVED** [review]

**Pull request merged**